platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,669 @@
+# Consensus Layer of the Consensus Node
+
+---
+
+## Summary
+
+Update the architecture for the consensus node to reduce complexity, improve performance, and improve stability.
+
+| Metadata           | Entities                                                 | 
+|--------------------|----------------------------------------------------------|
+| Designers          | Richard Bair, Jasper Potts, Oleg Mazurov, Austin Littley |
+| Functional Impacts | Consensus Node                                           |
+
+---
+
+## Purpose and Context
+
+Much of the motivation for this design can come down to paying down technical debt and simplifying the overall design.
+While the current design is full of amazing high quality solutions to various problems, it is also more complex than
+necessary, leading to hard-to-find or predict bugs, performance problems, or liveness (stability) issues while under
+load. This work is also necessary to prepare for autonomous node operation, and community nodes.
+
+1. This design defines several high-level modules providing strong encapsulation and isolation with strict contracts
+   between modules, leading to an overall simpler and more correct system.
+2. Assumptions and requirements that led to tight coupling between modules have been revisited, and where possible,
+   eliminated.
+3. The two fundamental modules are "consensus" and "execution". The Consensus module takes transactions and produces
+   rounds. Everything required to make that happen (gossip, event validation, hashgraph, event creation, etc.) is part
+   of the Consensus module. It is a library, and instances of the classes and interfaces within this library are created
+   and managed by the Execution module. The Consensus module does not persist state, has no main method, and has minimal
```

Comment:
The short answer is no.

The PCES is node-local. It is **not** part of the state. It is used by the node to recover from a node crash, but it is not used when we do a reconnect. So for example, if I have a node that is just trucking along, writing to the PCES, and then it shuts down for a while and starts back up, and finds it needs to reconnect, then it will reconnect and the PCES it had before is no longer useful. So as it starts going again, it starts writing new events to the PCES, and as it receives round information from the Hashgraph, it truncates the stored events from the "old" PCES.

This could (is?) also be optimized such that on reconnect we just chuck the PCES. It has discontinuity now, so we could aggressively dump it.

The only thing PCES does for us, is if we crash and restart, we can replay events to build up our hashgraph to where it should have been for the given round.

This is my understanding, @litt3 or @cody-littley correct me if I'm wrong here.

Reply:
I agree, Richard, I don't see any issues here. PCES is *not* state in the strict sense we use the word.

> This could (is?) also be optimized such that on reconnect we just chuck the PCES

Current cleanup is based on which states we have saved to disk, so the current system *will* chuck out old PCES after receiving a new state via reconnect and writing it to disk. We could do something similar every time the consensus module starts up in the new system, based on what the initial state is.

Reply:
If memory serves, we're still having to migrate PCES content when we do state transplants between networks?

Reply:
I could see transplant being useful to bring PCES, because basically, you're just restarting an existing network (not reconnecting a single node) and don't want to lose any transactions?

Reply:
@rbair23, I agree that the language here should updated to something like `The Consensus module does not persist state in the merkle tree` or add an asterisk and @edward-swirldslabs suggested explaining the role of PCES. The PCES is required on at least one node in the network for the network to startup without manual emergency intervention, so there is a case for it to be considered some kind of state even if not the merkle state.

Reply:
Indicating that consensus does not write to merkle state is sufficient to disambiguate for people who are not familiar with the common restriction of "state" to "merkle tree state" .

Reply:
I reworded the statement to be unambiguous.

```
The Consensus module does not persist state in the merkle tree, has no main method, and has minimal dependencies.
```